Add FeatureSelection tool

chrismayer opened the pull request:
This adds a FeatureSelection tool as described in #245.

![245-feature-select](https://user-images.githubusercontent.com/1185547/96707001-9a05e880-1397-11eb-8b18-e8bfe6fd88d8.gif)

chrismayer commented:
Some remarks: 

- Due to clustering of the vector features it sometimes looks like just one feature was clicked but in the grid two or more appear. 
- Would it be useful to inform the user if no feature was hit by the map click? Currently anything is done. Maybe a little feedback would increase usability?

simonseyock commented:
As long as your interaction is active you can disable the default behaviour by calling
```
olMap.set('enableDefaultClick', false);
```
and enable it again with setting it to true.

geographika commented on app/controller/button/FeatureSelectionButton.js:
```diff
@@ -0,0 +1,206 @@
+                menu: new Ext.menu.Menu({
+                    items: [
+                        {
+                            text: 'Add to selection',
```
Change to "Add to Selection"

chrismayer replied:
Done in 4f03a22

geographika commented on app/controller/button/FeatureSelectionButton.js:
```diff
@@ -0,0 +1,206 @@
+
+        if (!me.modeSelector) {
+            me.modeSelector = Ext.create('Ext.button.Split', {
+                text: 'Add to selection',
```
Change to "Add to Selection"

chrismayer replied:
Done in 4f03a22

geographika commented on app/controller/button/FeatureSelectionButton.js:
```diff
@@ -0,0 +1,206 @@
+                            }
+                        },
+                        {
+                            text: 'New selection',
```
Change to "New Selection"

chrismayer replied:
Done in 4f03a22

geographika reviewed (commented):
@chrismayer - playing with this now! Looks excellent. 
I'll need some time to try and test integration in the larger application.

A couple of initial comments. 

I might change the `selModel` on the grid to `MULTI` as then you can easily highlight all features in the grid on the map. Any foreseeable problems with this?

I think the clustering issue is fine.

In the old system when a feature was not selected I had an Ext animation flash a message in a status bar at the bottom of the screen. Maybe if an event was fired this functionality could be added in a future release.

chrismayer commented on app/controller/button/FeatureSelectionButton.js:
```diff
@@ -0,0 +1,206 @@
+        // collect all IDs of clicked features
+        me.map.forEachFeatureAtPixel(evt.pixel, function(feature, layer) {
+            // add check for correct layer
+            if (layer.id === view.queryLayer.id) {
```
Done in 4f03a22

chrismayer commented:
I addressed the review comments of @geographika.

Also removed the conflict with the GetFeatureInfo tool and added some behaviour if no feature was hit by clicking on the  map:

  - The view fires an event `'cmv-no-feature-selected'` 
  - By setting the views config option `showNoSelectionMessage` to `true` a little `Toast` is shown in the lower-right corner of the app with a message

chrismayer commented:
> I might change the selModel on the grid to MULTI as then you can easily highlight all features in the grid on the map. Any foreseeable problems with this?

I don't think so because this does not interact with the selection model at all.

geographika commented:
@chrismayer - I may make a follow-up pull request to allow a grid to have just the "Select by Polygon" or both select tools - but please merge this as-is - thanks.
